driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -181,90 +189,219 @@ public boolean isClosed() {
+        } catch (Exception e) {
+            close();
+            throw translateWriteException(e);
+        } finally {
+            writerLock.unlock();
         }
     }
 
     @Override
     public ResponseBuffers receiveMessage(final int responseTo) {
-        notNull("open", stream);
+        notNull("stream is open", stream);
         if (isClosed()) {
             throw new MongoSocketClosedException("Cannot read from a closed stream", getServerAddress());
-        } else {
-            while (!messages.containsKey(responseTo)) {
-                try {
-                    reading.acquire();
-                    if (messages.containsKey(responseTo)) {
-                        break;
-                    }
-                    ResponseBuffers responseBuffers = receiveResponseBuffers();
-                    try {
-                        connectionListener.messageReceived(new ConnectionMessageReceivedEvent(getId(),
-                                                                                              responseBuffers.getReplyHeader()
-                                                                                                             .getResponseTo(),
-                                                                                              responseBuffers.getReplyHeader()
-                                                                                                             .getMessageLength()));
-                    } catch (Throwable t) {
-                        LOGGER.warn("Exception when trying to signal messageReceived to the connectionListener", t);
-                    }
-                    messages.put(responseBuffers.getReplyHeader().getResponseTo(), new Response(responseBuffers, null));
-                } catch (Exception e) {
-                    close();
-                    messages.put(responseTo, new Response(null, translateReadException(e)));
-                } finally {
-                    reading.release();
-                }
-            }
-            Response response = messages.remove(responseTo);
-            if (response.hasError()) {
-                Throwable t = response.getError();
-                if (t instanceof MongoException) {
-                    throw (MongoException) t;
+        }
+
+        CountDownLatch localLatch = new CountDownLatch(1);
+        readerLock.lock();
+        try {
+            ResponseBuffers responseBuffers = receiveResponseBuffers();
+            messages.put(responseBuffers.getReplyHeader().getResponseTo(), responseBuffers);
+            readingPhase.getAndSet(localLatch).countDown();
+        } catch (Throwable t) {
+            exceptionThatPrecededStreamClosing = translateReadException(t);
+            close();
+            readingPhase.getAndSet(localLatch).countDown();
+        } finally {
+            readerLock.unlock();
```

Comment:
Unfortunately, no.  When I tried it there, FindBugs reported an error, saying that there is a path through which the readerLock is not unlocked (the path is where AtomicReference.getAndSet throws an exception.  which is not going to happen, but FindBugs doesn't know that.).